File: sdk-typescript/packages/notifications/src/notifications.ts
/**
 * WAVE SDK - Notifications API
 *
 * User notification preferences, delivery channels, and notification management.
 */

import type { WaveClient, PaginationParams, PaginatedResponse, Timestamps } from '@wave-av/core';

// ============================================================================
// Types
// ============================================================================

export type NotificationChannel = "in_app" | "email" | "push" | "sms" | "slack" | "webhook";
export type NotificationStatus = "unread" | "read" | "archived";
export type NotificationPriority = "low" | "normal" | "high" | "urgent";

export interface Notification extends Timestamps {
  id: string;
  user_id: string;
  type: string;
  title: string;
  body: string;
  status: NotificationStatus;
  priority: NotificationPriority;
  channel: NotificationChannel;
  action_url?: string;
  metadata?: Record<string, unknown>;
  read_at?: string;
}

export interface NotificationPreferences {
  user_id: string;
  channels: Record<NotificationChannel, boolean>;
  categories: Record<string, { enabled: boolean; channels: NotificationChannel[] }>;
  quiet_hours?: { start: string; end: string; timezone: string };
  digest_frequency?: "realtime" | "hourly" | "daily" | "weekly";
}

export interface ListNotificationsParams extends PaginationParams {
  status?: NotificationStatus;
  type?: string;
  priority?: NotificationPriority;
  channel?: NotificationChannel;
  since?: string;
}

// ============================================================================
// Notifications API
// ============================================================================

/**
 * User notification management: preferences, channels, read state, and delivery.
 *
 * @example
 * ```typescript
 * const notifications = await wave.notifications.list({ status: 'unread' });
 * await wave.notifications.markAsRead(notifications.data[0].id);
 * await wave.notifications.markAllRead();
 * const prefs = await wave.notifications.getPreferences();
 * ```
 */
export class NotificationsAPI {
  private readonly client: WaveClient;
  private readonly basePath = "/v1/notifications";

  constructor(client: WaveClient) {
    this.client = client;
  }

  /** List notifications with optional filters. */
  async list(params?: ListNotificationsParams): Promise<PaginatedResponse<Notification>> {
    return this.client.get<PaginatedResponse<Notification>>(this.basePath, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }

  /** Get a single notification by ID. */
  async get(notificationId: string): Promise<Notification> {
    return this.client.get<Notification>(`${this.basePath}/${notificationId}`);
  }

  /** Mark a notification as read. */
  async markAsRead(notificationId: string): Promise<Notification> {
    return this.client.post<Notification>(`${this.basePath}/${notificationId}/read`);
  }

  /** Mark all notifications as read. */
  async markAllRead(): Promise<{ updated: number }> {
    return this.client.post<{ updated: number }>(`${this.basePath}/mark-all-read`);
  }

  /** Archive a notification. */
  async archive(notificationId: string): Promise<Notification> {
    return this.client.post<Notification>(`${this.basePath}/${notificationId}/archive`);
  }

  /** Delete a notification. */
  async remove(notificationId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/${notificationId}`);
  }

  /** Get unread count. */
  async getUnreadCount(): Promise<{ count: number }> {
    return this.client.get<{ count: number }>(`${this.basePath}/unread-count`);
  }

  /** Get notification preferences. */
  async getPreferences(): Promise<NotificationPreferences> {
    return this.client.get<NotificationPreferences>(`${this.basePath}/preferences`);
  }

  /** Update notification preferences. */
  async updatePreferences(
    preferences: Partial<NotificationPreferences>,
  ): Promise<NotificationPreferences> {
    return this.client.patch<NotificationPreferences>(`${this.basePath}/preferences`, preferences);
  }
}

export function createNotificationsAPI(client: WaveClient): NotificationsAPI {
  return new NotificationsAPI(client);
}

File: sdk-python/wave/notifications.py
"""WAVE SDK - Notifications API. User notifications, preferences, and delivery channels."""
from __future__ import annotations
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class Notification(BaseModel):
    id: str; user_id: str; type: str; title: str; body: str; status: str; priority: str; channel: str; action_url: str | None = None; read_at: str | None = None; created_at: str; updated_at: str

class NotificationsAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/notifications"
    def list(self, **params: Any) -> dict: return self._client.get(self._base, params={k: v for k, v in params.items() if v is not None})
    def get(self, notification_id: str) -> Notification: return Notification(**self._client.get(f"{self._base}/{notification_id}"))
    def mark_as_read(self, notification_id: str) -> Notification: return Notification(**self._client.post(f"{self._base}/{notification_id}/read"))
    def mark_all_read(self) -> dict: return self._client.post(f"{self._base}/mark-all-read")
    def archive(self, notification_id: str) -> Notification: return Notification(**self._client.post(f"{self._base}/{notification_id}/archive"))
    def remove(self, notification_id: str) -> None: self._client.delete(f"{self._base}/{notification_id}")
    def get_unread_count(self) -> dict: return self._client.get(f"{self._base}/unread-count")
    def get_preferences(self) -> dict: return self._client.get(f"{self._base}/preferences")
    def update_preferences(self, **prefs: Any) -> dict: return self._client.patch(f"{self._base}/preferences", json=prefs)
